File: Consistent Hashing II.cc
class Solution {
public:
    // @param n a positive integer
    // @param k a positive integer
    // @return a Solution object
    template<typename T>
    void swap(T &a, T &b)
    {
        T t = a;
        a = b;
        b = t;
    }

    static Solution create(int n, int k) {
        // Write your code here
        int nn = n;
        int kk = k;
        vector<int> v = {};
        for(auto i = 0; i < nn; i++){
            v.push_back(i);
        }
        for(auto i = 0; i < nn; i++){
            int r = i + std::rand() % (nn - i + 1);
            swap(v[i], v[r]);
        }
        unordered_map<int, int> h;
    }

    // @param machine_id an integer
    // @return a list of shard ids
    vector<int> addMachine(int machine_id) {
        // Write your code here
        vector<int> res = {};
        for (auto i = 0; i < kk; i++){
            res.push_back(v.pop_back());
        }
        for(auto n : res)
        {
            h[n] = machine_id;
        }
        return res;
    }

    // @param hashcode an integer
    // @return a machine id
    int getMachineIdByHashCode(int hashcode) {
        // Write your code here
        while (h.find(hashcode) == h.end()){
            hashcode ++;
        }
        return h[hashcode];
    }
};

File: Consistent Hashing II.py
import random
class Solution:

    # @param {int} n a positive integer
    # @param {int} k a positive integer
    # @return {Solution} a Solution object
    @classmethod
    def create(cls, n, k):
        # Write your code here
        cls.n = n
        cls.k = k
        cls.valuelist = [i for i in xrange(cls.n)]
        cls.h = {}
        return cls


    # @param {int} machine_id an integer
    # @return {int[]} a list of shard ids
    def addMachine(self, machine_id):
        # write your code here
        res = random.sample(self.valuelist,  self.k)
        for e in res:
            self.h[e] = machine_id
            self.valuelist.remove(e)
        return res


    # @param {int} hashcode an integer
    # @return {int} a machine id
    def getMachineIdByHashCode(self, hashcode):
        # write your code here
        while hashcode not in self.h:
            hashcode += 1
        return self.h[hashcode]
a = Solution()
print a.create(100,3)
print a.addMachine(1)
print a.getMachineIdByHashCode(4)
print a.addMachine(2)
print a.getMachineIdByHashCode(61)
print a.getMachineIdByHashCode(91)
